File: apps/memory-lab-py/memory_lab_py/tools.py
from __future__ import annotations

from typing import Any


def _to_limit(raw: Any, fallback: int) -> int:
    if isinstance(raw, int) and raw > 0:
        return raw
    if isinstance(raw, float) and raw > 0:
        return int(raw)
    if isinstance(raw, str) and raw.strip():
        try:
            parsed = int(raw.strip())
            return max(1, parsed)
        except ValueError:
            return fallback
    return fallback


class MemoryTools:
    def __init__(self, repository, retriever):
        self.repository = repository
        self.retriever = retriever

    def memory_recall(self, query: str, limit: int = 6, include_facts: bool = True, with_trace: bool = True) -> dict:
        if not query.strip():
            return {"ok": False, "error": "query is required"}
        result = self.retriever.retrieve(
            query,
            {
                "l2Limit": _to_limit(limit, 6),
                "l1Limit": _to_limit(limit, 6),
                "l0Limit": max(3, int(_to_limit(limit, 6) / 2)),
                "includeFacts": include_facts,
                "withTrace": with_trace,
            },
        )
        return {"ok": True, **result}

    def memory_store(self, content: str, fact_key: str | None = None, confidence: float = 0.8) -> dict:
        text = content.strip()
        if not text:
            return {"ok": False, "error": "content is required"}
        explicit_key = fact_key.strip() if isinstance(fact_key, str) and fact_key.strip() else f"manual:{text[:64].lower()}"
        bounded_confidence = max(0.1, min(1.0, float(confidence)))
        fact = {
            "factKey": explicit_key,
            "factValue": text,
            "confidence": bounded_confidence,
        }
        self.repository.upsert_global_facts([fact])
        return {"ok": True, "stored": fact}

    def search_l2(self, query: str, search_type: str = "general", limit: int = 8) -> dict:
        intent = search_type if search_type in ("time", "project") else "general"
        results = self.retriever.search_l2(query, intent, _to_limit(limit, 8))
        return {"ok": True, "count": len(results), "results": results}

    def search_l1(self, query: str, l1_ids: list[str] | None = None, limit: int = 8) -> dict:
        ids = l1_ids or []
        results = self.retriever.search_l1(query, ids, _to_limit(limit, 8))
        return {"ok": True, "count": len(results), "results": results}

    def search_l0(self, query: str, l0_ids: list[str] | None = None, limit: int = 6) -> dict:
        ids = l0_ids or []
        results = self.retriever.search_l0(query, ids, _to_limit(limit, 6))
        return {"ok": True, "count": len(results), "results": results}

    def memory_search(self, query: str, limit: int = 6) -> dict:
        result = self.memory_recall(query=query, limit=limit, include_facts=True, with_trace=True)
        if not result.get("ok"):
            return result
        result["alias"] = "memory_recall"
        return result

    def invoke(self, tool_name: str, params: dict[str, Any] | None = None) -> dict:
        payload = params or {}
        if tool_name == "memory_recall":
            return self.memory_recall(
                query=str(payload.get("query", "")),
                limit=_to_limit(payload.get("limit"), 6),
                include_facts=bool(payload.get("includeFacts", True)),
                with_trace=bool(payload.get("withTrace", True)),
            )
        if tool_name == "memory_store":
            return self.memory_store(
                content=str(payload.get("content", "")),
                fact_key=payload.get("factKey"),
                confidence=float(payload.get("confidence", 0.8)),
            )
        if tool_name == "search_l2":
            return self.search_l2(
                query=str(payload.get("query", "")),
                search_type=str(payload.get("type", "general")),
                limit=_to_limit(payload.get("limit"), 8),
            )
        if tool_name == "search_l1":
            raw_ids = payload.get("l1Ids", [])
            l1_ids = [str(item) for item in raw_ids] if isinstance(raw_ids, list) else []
            return self.search_l1(
                query=str(payload.get("query", "")),
                l1_ids=l1_ids,
                limit=_to_limit(payload.get("limit"), 8),
            )
        if tool_name == "search_l0":
            raw_ids = payload.get("l0Ids", [])
            l0_ids = [str(item) for item in raw_ids] if isinstance(raw_ids, list) else []
            return self.search_l0(
                query=str(payload.get("query", "")),
                l0_ids=l0_ids,
                limit=_to_limit(payload.get("limit"), 6),
            )
        if tool_name == "memory_search":
            return self.memory_search(
                query=str(payload.get("query", "")),
                limit=_to_limit(payload.get("limit"), 6),
            )
        return {"ok": False, "error": f"unknown tool: {tool_name}"}

File: packages/openclaw-memory-plugin/src/tools.ts
import { MemoryRepository, ReasoningRetriever } from "./core/index.js";
import type { PluginTool } from "./plugin-api.js";

function jsonResult(payload: unknown): string {
  try {
    return JSON.stringify(payload, null, 2);
  } catch {
    return String(payload);
  }
}

function toLimit(raw: unknown, fallback: number): number {
  if (typeof raw === "number" && Number.isFinite(raw)) return Math.max(1, Math.floor(raw));
  if (typeof raw === "string" && raw.trim()) {
    const parsed = Number.parseInt(raw, 10);
    if (Number.isFinite(parsed)) return Math.max(1, parsed);
  }
  return fallback;
}

export function buildPluginTools(
  repository: MemoryRepository,
  retriever: ReasoningRetriever,
): PluginTool[] {
  return [
    {
      name: "memory_recall",
      label: "Recall Multi-Level Memory",
      description: "Retrieve memory context using L2->L1->L0 fallback reasoning.",
      parameters: {
        type: "object",
        properties: {
          query: { type: "string", description: "Question or topic to search in memory." },
          limit: { type: "number", description: "Maximum items per level." },
        },
        required: ["query"],
      },
      async execute(_id, params) {
        const input = (params ?? {}) as Record<string, unknown>;
        const query = typeof input.query === "string" ? input.query : "";
        if (!query.trim()) {
          return jsonResult({ ok: false, error: "query is required" });
        }
        const limit = toLimit(input.limit, 6);
        const result = await retriever.retrieve(query, {
          retrievalMode: "explicit",
          l2Limit: limit,
          l1Limit: limit,
          l0Limit: Math.max(3, Math.floor(limit / 2)),
        });
        return jsonResult({ ok: true, ...result });
      },
    },
    {
      name: "memory_store",
      label: "Append Profile Note",
      description: "Append a manual note into the global user profile summary.",
      parameters: {
        type: "object",
        properties: {
          content: { type: "string", description: "Profile note content to append." },
        },
        required: ["content"],
      },
      async execute(_id, params) {
        const input = (params ?? {}) as Record<string, unknown>;
        const content = typeof input.content === "string" ? input.content.trim() : "";
        if (!content) return jsonResult({ ok: false, error: "content is required" });
        const profile = repository.appendToGlobalProfile(content);
        return jsonResult({ ok: true, profile });
      },
    },
    {
      name: "search_l2",
      label: "Search L2 Index",
      description: "Search time/project second-level indexes by fuzzy query.",
      parameters: {
        type: "object",
        properties: {
          query: { type: "string" },
          type: { type: "string", enum: ["time", "project", "general"] },
          limit: { type: "number" },
        },
        required: ["query"],
      },
      async execute(_id, params) {
        const input = (params ?? {}) as Record<string, unknown>;
        const query = typeof input.query === "string" ? input.query : "";
        const limit = toLimit(input.limit, 8);
        const type = typeof input.type === "string" ? input.type : "general";
        const intent = type === "time" || type === "project" ? type : "general";
        const results = await retriever.searchL2(query, intent, limit);
        return jsonResult({ ok: true, count: results.length, results });
      },
    },
    {
      name: "search_l1",
      label: "Search L1 Index",
      description: "Search structured conversation windows.",
      parameters: {
        type: "object",
        properties: {
          query: { type: "string" },
          l1Ids: { type: "array", items: { type: "string" } },
          limit: { type: "number" },
        },
        required: ["query"],
      },
      async execute(_id, params) {
        const input = (params ?? {}) as Record<string, unknown>;
        const query = typeof input.query === "string" ? input.query : "";
        const ids = Array.isArray(input.l1Ids) ? input.l1Ids.filter((v): v is string => typeof v === "string") : [];
        const limit = toLimit(input.limit, 8);
        const results = await retriever.searchL1(query, ids, limit);
        return jsonResult({ ok: true, count: results.length, results });
      },
    },
    {
      name: "search_l0",
      label: "Search L0 Logs",
      description: "Search raw conversation logs as final fallback.",
      parameters: {
        type: "object",
        properties: {
          query: { type: "string" },
          l0Ids: { type: "array", items: { type: "string" } },
          limit: { type: "number" },
        },
        required: ["query"],
      },
      async execute(_id, params) {
        const input = (params ?? {}) as Record<string, unknown>;
        const query = typeof input.query === "string" ? input.query : "";
        const ids = Array.isArray(input.l0Ids) ? input.l0Ids.filter((v): v is string => typeof v === "string") : [];
        const limit = toLimit(input.limit, 6);
        const results = await retriever.searchL0(query, ids, limit);
        return jsonResult({ ok: true, count: results.length, results });
      },
    },
  ];
}
